fixed dot operator recursive loop & macro suggest

saem opened the pull request.

Araq committed:
basic stability improvements; refs nimsuggest

saem committed:
fixed dot operator recursive loop & macro suggest

metagn commented on compiler/semexprs.nim:
```diff
@@ -1422,7 +1422,10 @@ proc dotTransformation(c: PContext, n: PNode): PNode =
-    result.flags.incl nfDotField
+    # we only want to do the dotField transform once per node, so the new node
+    # should not keep the nfDotField flag, as it's the result of a transform.
+    if nfDotField notin n.flags:
+      result.flags.incl nfDotField
```
```suggestion
    if nfDotField notin n.flags:
      result.flags.incl nfDotField
    else:
      result.flags.excl nfDotField
```

Is this what was meant here? Because the current behavior is no different from the previous behavior

saem replied:
No, I was tired and it should not include the flag as it was in the node before. The transform has already happened and it should be "consumed".

timotheecour commented:
why is the targetted branch `nim-lang:araq-stability-1` instead of `devel`?

saem commented:
> why is the targetted branch `nim-lang:araq-stability-1` instead of `devel`?

Because Araq started it so I'm integrating with him and any other fixes but if he wants me to retarget devel I can do that too. Credit where credit is due and all that.

saem committed:
hacky fix for run away dot operator sem check

Committing this mostly to make the issue more clear. Perhaps get better
feedback.

timotheecour commented on compiler/semexprs.nim:
```diff
@@ -1419,6 +1419,12 @@ proc dotTransformation(c: PContext, n: PNode): PNode =
     result.add n[1]
     result.add copyTree(n[0])
+  elif n[0].typ != nil and n[0].typ.kind == tyError:
+    # XXX: this hack avoids the issue where we keep reevaluating the same
```
a test linked to here is worth a thousand words (but ok to explain too)

saem replied:
It's t11166.nim that's demonstrates the issue, the reason I didn't reference it in the comment is because I believe (could be wrong) that the issue should be fixed elsewhere.

timotheecour replied:
still, it helps to add refs t11166.nim in the comment

timotheecour commented:
CI failure is related to https://github.com/nim-lang/Nim/issues/16956

saem committed:
semExprWithType seems like a better place to check

saem commented:
> CI failure is related to #16956

I think this is an issue with how AST traversal is likely different in this branch, likely uncovered more issues to fix.

saem committed:
some progress

timotheecour commented on tests/js/t11166.nim:
```diff
@@ -1,4 +1,19 @@
+    props: int
+
+var c: C
+
```
just use:
```nim
doAssert compiles(c.props)
doAssert not copmiles(d.props)
```

saem committed:
fixed error messages const case expressions

saem committed:
Clean-up test

saem commented:
I can't replicate the test failures on my machine ~and I can't seem to get particularly useful output from the CI logs or I'm not looking in the right place~. 😕 

But I've found a much bigger problem looking at the JS output, back to the drawing board.

saem committed:
stopped the dot operator madness

No longer get infinite recursion when seming broken code with a dot
operator macro like in jsffi.

saem commented:
Wow, it passed -- I'd almost given up hope.

I suppose I should spruce up the PR/commit message for the squashening? 🤔
